doc/man/tst.3
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "tst 3"
.TH tst 3 "2008-04-06" "INN 2.4.6" "InterNetNews Documentation"
.SH "NAME"
tst \- ternary search trie functions
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fB#include <inn/tst.h>\fR
.PP
\&\fBstruct tst;\fR
.PP
\&\fBstruct tst *tst_init(int \fR\fInode_line_width\fR\fB);\fR
.PP
\&\fBvoid tst_cleanup(struct tst *\fR\fItst\fR\fB);\fR
.PP
\&\fBint tst_insert(struct tst *\fR\fItst\fR\fB, const unsigned char *\fR\fIkey\fR\fB, void *\fR\fIdata\fR\fB, int \fR\fIoption\fR\fB, void **\fR\fIexist_ptr\fR\fB);\fR
.PP
\&\fBvoid *tst_search(struct tst *\fR\fItst\fR\fB, const unsigned char *\fR\fIkey\fR\fB);\fR
.PP
\&\fBvoid *tst_delete(struct tst *\fR\fItst\fR\fB, const unsigned char *\fR\fIkey\fR\fB);\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBtst_init\fR allocates memory for members of \fIstruct tst\fR, and
allocates the first \fInode_line_width\fR nodes. A \s-1NULL\s0 pointer is
returned by \fBtst_init\fR if any part of the memory allocation fails. On
success, a pointer to a \fIstruct tst\fR is returned.
.PP
The value for \fInode_line_width\fR must be chosen very carefully. One
node is required for every character in the tree. If you choose a
value that is too small, your application will spend too much time
calling \fImalloc\fR\|(3) and your node space will be too spread out. Too large
a value is just a waste of space.
.PP
\&\fBtst_cleanup\fR frees all memory allocated to nodes, internal structures,
as well as \fItst\fR itself.
.PP
\&\fBtst_insert\fR inserts the string \fIkey\fR into the tree. Behavior when a
duplicate key is inserted is controlled by \fIoption\fR. If \fIkey\fR is
already in the tree then \fB\s-1TST_DUPLICATE_KEY\s0\fR is returned, and the
data pointer for the existing key is placed in \fIexist_ptr\fR.  If
\&\fIoption\fR is set to \fB\s-1TST_REPLACE\s0\fR then the existing data pointer for
the existing key is replaced by \fIdata\fR.  Note that the old data
pointer will still be placed in \fIexist_ptr\fR.
.PP
If a duplicate key is encountered and \fIoption\fR is not set to
\&\fB\s-1TST_REPLACE\s0\fR then \fB\s-1TST_DUPLICATE_KEY\s0\fR is returned. If \fIkey\fR is
zero length then \fB\s-1TST_NULL_KEY\s0\fR is returned. A successful insert or
replace returns \fB\s-1TST_OK\s0\fR. A return value of \fB\s-1TST_ERROR\s0\fR indicates
that a memory allocation error occurred while trying to grow the node
free.
.PP
Note that the \fIdata\fR argument must never be \fB\s-1NULL\s0\fR. If it is, then
calls to \fBtst_search\fR will fail for a key that exists because the
data value was set to \fB\s-1NULL\s0\fR, which is what \fBtst_search\fR returns. If
you just want a simple existence tree, use the \fBtst\fR pointer as the
data pointer.
.PP
\&\fBtst_search\fR finds the string \fIkey\fR in the tree if it exists and
returns the data pointer associated with that key.
.PP
If \fIkey\fR is not found then \fB\s-1NULL\s0\fR is returned, otherwise the data pointer
associated with \fIkey\fR is returned.
.PP
\&\fBtst_delete\fR deletes the string \fIkey\fR from the tree if it exists and
returns the data pointer assocaited with that key.
.PP
If \fIkey\fR is not found then \fB\s-1NULL\s0\fR is returned, otherwise the data
pointer associated with \fIkey\fR is returned.
.SH "HISTORY"
.IX Header "HISTORY"
Converted to \s-1POD\s0 from Peter A. Friend's ternary search trie
documentation by Alex Kiernan <alex.kiernan@thus.net> for InterNetNews
2.4.0.
.PP
$Id$


NEWS
Changes in 2.4.6

    * Fixed the segfault of the radius authenticator when none of the radius
      servers respond.  Thanks to Matija Nalis for this patch.

    * Fixed a lost initialization in buffindexed, which resolves a potential
      segfault, thanks to a patch by Kirill Berezin.

    * INN now properly supports Perl 5.10.0; Perl filters were causing innd
      to segfault on a few systems like FreeBSD.

    * Fixed a long-standing bug which affected Perl hooks for innd: the
      variable containing the body of an article was not properly created,
      which caused regular expressions matching new lines to fail.  It
      especially affected filters like Cleanfeed which sometimes failed to
      detect unwanted articles.

      To fix that issue, Julien Elie added the use of a shared string,
      available since Perl 5.7.2, with a fall back to a slower but reliable
      copy of such bodies in case the function is not available.  Using a
      Perl version superior to 5.7.2 is therefore recommended.

    * Fixed two bugs which could prevent nnrpd from being run as a daemon in
      FreeBSD.  Thanks to Johan van Selst for having identified the problem
      and to Kai Gallasch for having provided a testing FreeBSD server.  The
      listening address was not initialized to "::0" or 0.0.0.0 when the -b
      flag was not used and an incorrect size was given when IPv6 was
      enabled and the binding done using IPv4.

    * Some annoying assertion failures occurring in innfeed have been fixed
      by Russ Allbery and Julien Elie.

    * Fixed a bug in mod-active for aliased newsgroups.  Only "=" was
      written to the active file.  Thanks to D. Stussy for this patch.

    * Fixed a bug which caused innd not to honour the Ad flag in newsfeeds.

    * Fixed a bug in the IP address displayed for "localhost" in innd's
      status file.  It was not correctly initialized.

    * Fixed a permission issue:  XHDR and XPAT were not checking the rights
      the user had to read articles when accessing them by their message-ID.

    * Fixed a bug in the replies of XHDR, XOVER and XPAT when the newsgroup
      is empty.  Two initial replies were sent instead of one:  the right
      420 code followed by a wrong 224 code.

    * When no newsgroup is selected, LISTGROUP now returns the right 412
      code (instead of 481).

    * inncheck now uses a range of permissions to see whether the file modes
      are correctly set.  Therefore, different configurations depending on
      the security the user wants to enforce on his sytem are possible.

    * A new improved version of docheckgroups is shipped with INN.  The -u
      flag permits to automatically update the newsgroups file (with a
      proper number of tabulations and an alphabetical sort), removing
      obsolete descriptions and adding new ones.  A second argument on
      command-line permits to specify which newsgroups should not be
      checked, so as not to treat them.

    * An *email=* keyword has been added by James Ralston to news.daily in
      order to supply another mail address than the one set at configure
      time for Usenet daily reports.

    * An updated moderators file with information about the aioe.*, perl.*
      and si.* hierarchies is provided; control.ctl is also up to date.

    * INN supports Berkeley DB 4.7, which is the recommended version to use
      owing to various bugs affecting previous versions of Berkeley DB.

    * Other minor bugs have also been fixed.

Changes in 2.4.5

    * Fixed the "alarm signal" around "SSL_read" in nnrpd:  it allows a
      proper disconnection of news clients which were previously hanging
      when posting an article through a SSL connection.  Moreover, the
      *clienttimeout* parameter now works on SSL connections.  Thanks to
      Matija Nalis for the patch.

    * SO_KEEPALIVE is now implemented for SSL TCP connections on systems
      which support it, allowing system detection and closing the dead TCP
      SSL connections automatically after system-specified time.  Thanks to
      Matija Nalis for the patch.

    * Fixed a segmentation fault when an article of a size greater than
      remaining stack is retrieved via SSL.  Thanks to Chris Caputo for this
      patch.

    * Fixed a few segfaults and bugs which affected both Python innd and
      nnrpd hooks.  They no longer check the existence of methods not used
      by the hooked script.  An issue with Python exception handling was
      also fixed, as well as a segfault fixed by Russ Allbery which happened
      whenever one closes and then reopens Python in the same process. 
      Julien Elie also fixed a bug when reloading Python filters (they were
      not always correctly reloaded) and a segfault when generating access
      groups with embedded Python filters for nnrpd.  Many thanks to David
      Hlacik for its bug reports.

    * The nnrpd.py stub file in order to test Python nnrpd hooks, as
      mentioned in their documentation, is now installed; only INN.py was
      previously installed in *pathfilter*.  Also fixed a bug in INN.py and
      add missing methods to it.

    * Fixed a long-standing bug in innreport which prevented it from
      correctly reporting nnrpd and innfeed log messages.

    * Fixed a hang in Perl hooks on (at least) HP/PA since Perl 5.10.

    * Fixed a compilation problem on some platforms because of AF_INET6
      which was not inside a HAVE_INET6 block in innfeed.

    * Fixed a bug in innfeed which contained thrice the same IPs for each
      peer; it unnecessarily slowed the peer IP rotation for innfeed. 
      Thanks, D. Stussy, for having seen that.  Miquel van Smoorenburg
      provided the patch.

    * A new *heavily* improved version of pullnews is shipped with this INN
      release.  This new version is provided by Geraint Edwards.  He added
      no more than 16 flags, fixed some bugs and integrated the backupfeed

lib/tst.c
/*  $Id$
**
**  Ternary search trie implementation.
**
**  A ternary search trie stores key/value pairs where the key is a
**  nul-terminated string and the value is an arbitrary pointer.  It uses a
**  data structure designed for fast lookups, where each level of the trie
**  represents a character in the string being searched for.
**
**  This implementation is based on the implementation by Peter A. Friend
**  (version 1.3), but has been assimilated into INN and modified to use INN
**  formatting conventions.  If new versions are released, examine the
**  differences between that version and version 1.3 (which was checked into
**  INN as individual files in case it's no longer available) and then apply
**  the changes to this file.
**
**  Copyright (c) 2002, Peter A. Friend
**  All rights reserved.
**
**  Redistribution and use in source and binary forms, with or without
**  modification, are permitted provided that the following conditions are
**  met:
**
**  Redistributions of source code must retain the above copyright notice,
**  this list of conditions and the following disclaimer.
**
**  Redistributions in binary form must reproduce the above copyright notice,
**  this list of conditions and the following disclaimer in the documentation
**  and/or other materials provided with the distribution.
**
**  Neither the name of Peter A. Friend nor the names of his contributors may
**  be used to endorse or promote products derived from this software without
**  specific prior written permission.
**
**  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
**  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
**  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
**  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
**  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
**  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
**  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
**  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
**  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
**  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
**  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include "config.h"
#include "clibrary.h"

#include "inn/tst.h"
#include "inn/libinn.h"


/* A single node in the ternary search trie.  Stores a character, which is
   part of the string formed by walking the tree from its root down to the
   node, and left, right, and middle pointers to child nodes.  If value is
   non-zero (not a nul), middle is the pointer to follow if the desired
   string's character matches value.  left is used if it's less than value and
   right is used if it's greater than value.  If value is zero, this is a
   terminal node, and middle holds a pointer to the data associated with the
   string that ends at this point. */
struct node {
    unsigned char value;
    struct node *left;
    struct node *middle;
    struct node *right;
};

/* The search trie structure.  node_line_width is the number of nodes that are
   allocated at a time, and node_lines holds a linked list of groups of nodes.
   The free_list is a linked list (through the middle pointers) of available
   nodes to use, and head holds pointers to the first nodes for each possible
   first letter of the string. */
struct tst {
    int node_line_width;
    struct node_lines *node_lines;
    struct node *free_list;
    struct node *head[256];
};

/* A simple linked list structure used to hold all the groups of nodes. */
struct node_lines {
    struct node *node_line;
    struct node_lines *next;
};


/*
**  Given a node and a character, decide whether a new node for that character
**  should be placed as the left child of that node.  (If false, it should be
**  placed as the right child.)
*/
#define LEFTP(n, c) (((n)->value == 0) ? ((c) < 64) : ((c) < (n)->value))


/*
**  Allocate new nodes for the free list, called when the free list is empty.
*/
static void
tst_grow_node_free_list(struct tst *tst)
{
    struct node *current_node;
    struct node_lines *new_line;
    int i;

    new_line = xmalloc(sizeof(struct node_lines));
    new_line->node_line = xcalloc(tst->node_line_width, sizeof(struct node));
    new_line->next = tst->node_lines;
    tst->node_lines = new_line;

    current_node = tst->node_lines->node_line;
    tst->free_list = current_node;
    for (i = 1; i < tst->node_line_width; i++) {
        current_node->middle = &(tst->node_lines->node_line[i]);
        current_node = current_node->middle;
    }
    current_node->middle = NULL;
}


/*
**  Grab a node from the free list and initialize it with the given value.
*/
static struct node *
tst_get_free_node(struct tst *tst, unsigned char value)
{
    struct node *free_node;

    if (tst->free_list == NULL)
        tst_grow_node_free_list(tst);
    free_node = tst->free_list;
    tst->free_list = tst->free_list->middle;
    free_node->middle = NULL;
    free_node->value = value;
    return free_node;
}


/*
**  tst_init allocates memory for members of struct tst, and allocates the
**  first node_line_width nodes.  The value for width must be chosen very
**  carefully.  One node is required for every character in the tree.  If you
**  choose a value that is too small, your application will spend too much
**  time calling malloc and your node space will be too spread out.  Too large
**  a value is just a waste of space.
*/
struct tst *
tst_init(int width)
{
    struct tst *tst;

    tst = xcalloc(1, sizeof(struct tst));
    tst->node_lines = NULL;
    tst->node_line_width = width;
    tst_grow_node_free_list(tst);
    return tst;
}


/*
**  tst_insert inserts the string key into the tree.  Behavior when a
**  duplicate key is inserted is controlled by option.  If key is already in
**  the tree then TST_DUPLICATE_KEY is returned, and the data pointer for the
**  existing key is placed in exist_ptr.  If option is set to TST_REPLACE then
**  the existing data pointer for the existing key is replaced by data.  The
**  old data pointer will still be placed in exist_ptr.
**
**  If a duplicate key is encountered and option is not set to TST_REPLACE
**  then TST_DUPLICATE_KEY is returned.  If key is zero-length, then
**  TST_NULL_KEY is returned.  A successful insert or replace returns TST_OK.
**
**  The data argument may not be NULL; if it is, TST_NULL_DATA is returned.
**  If you just want a simple existence tree, use the tst pointer as the data
**  pointer.
*/
int
tst_insert(struct tst *tst, const unsigned char *key, void *data, int option,
           void **exist_ptr)
{
    struct node *current_node = NULL;
    struct node **root_node = NULL;
    int key_index;

    if (data == NULL)
        return TST_NULL_DATA;

    if (key == NULL || *key == '\0')
        return TST_NULL_KEY;

    key_index = 1;
    if (tst->head[*key] == NULL)
        root_node = &tst->head[*key];
    else
        current_node = tst->head[*key];

    while (root_node == NULL) {
        if (key[key_index] == current_node->value) {
            if (key[key_index] == '\0') {
                if (exist_ptr != NULL)
                    *exist_ptr = current_node->middle;
                if (option == TST_REPLACE) {
                    current_node->middle = data;
                    return TST_OK;
                } else
                    return TST_DUPLICATE_KEY;
            }
            if (current_node->middle == NULL)
                root_node = &current_node->middle;
            else {
                current_node = current_node->middle;
                key_index++;
            }
        } else if (LEFTP(current_node, key[key_index])) {
            if (current_node->left == NULL)
                root_node = &current_node->left;
            else
                current_node = current_node->left;
        } else {
            if (current_node->right == NULL)
                root_node = &current_node->right;
            else
                current_node = current_node->right;
        }

    }

    *root_node = tst_get_free_node(tst, key[key_index]);
    current_node = *root_node;

    while (key[key_index] != '\0') {
        key_index++;
        current_node->middle = tst_get_free_node(tst, key[key_index]);
        current_node = current_node->middle;
    }

    current_node->middle = data;
    return TST_OK;
}


/*
**  tst_search finds the string key in the tree if it exists and returns the
**  data pointer associated with that key or NULL if it's not found.
*/
void *
tst_search(struct tst *tst, const unsigned char *key)
{
    struct node *current_node;
    int key_index;

    if (key == NULL || *key == '\0')
        return NULL;

    if (tst->head[*key] == NULL)
        return NULL;

    current_node = tst->head[*key];
    key_index = 1;
    while (current_node != NULL) {
        if (key[key_index] == current_node->value) {
            if (current_node->value == '\0')
                return current_node->middle;
            else {
                current_node = current_node->middle;
                key_index++;
                continue;
            }
        } else if (LEFTP(current_node, key[key_index]))
            current_node = current_node->left;
        else
            current_node = current_node->right;
    }
    return NULL;
}


/*
**  tst_delete deletes the string key from the tree if it exists and returns
**  the data pointer assocaited with that key, or NULL if it wasn't found.
*/
void *
tst_delete(struct tst *tst, const unsigned char *key)
{
    struct node *current_node;
    struct node *current_node_parent;
    struct node *last_branch;
    struct node *last_branch_parent;
    struct node *next_node;
    struct node *last_branch_replacement;
    struct node *last_branch_dangling_child;
    int key_index;

    if (key == NULL || *key == '\0')
        return NULL;

    if (tst->head[*key] == NULL)
        return NULL;

    last_branch = NULL;
    last_branch_parent = NULL;
    current_node = tst->head[*key];
    current_node_parent = NULL;
    key_index = 1;
    while (current_node != NULL) {
        if (key[key_index] == current_node->value) {
            if (current_node->left != NULL || current_node->right != NULL) {
                last_branch = current_node;
                last_branch_parent = current_node_parent;
            }
            if (key[key_index] == '\0')
                break;
            else {
                current_node_parent = current_node;
                current_node = current_node->middle;
                key_index++;
            }
        } else if (LEFTP(current_node, key[key_index])) {
            last_branch_parent = current_node;
            current_node_parent = current_node;
            current_node = current_node->left;
            last_branch = current_node;
        } else {
            last_branch_parent = current_node;
            current_node_parent = current_node;
            current_node = current_node->right;
            last_branch = current_node;
        }
    }
    if (current_node == NULL)
        return NULL;

    if (last_branch == NULL) {
        next_node = tst->head[*key];
        tst->head[*key] = NULL;
    } else if (last_branch->left == NULL && last_branch->right == NULL) {
        if (last_branch_parent->left == last_branch)
            last_branch_parent->left = NULL;
        else
            last_branch_parent->right = NULL;
        next_node = last_branch;
    } else {
        if (last_branch->left != NULL && last_branch->right != NULL) {
            last_branch_replacement = last_branch->right;
            last_branch_dangling_child = last_branch->left;
        } else if (last_branch->right != NULL) {
            last_branch_replacement = last_branch->right;
            last_branch_dangling_child = NULL;
        } else {
            last_branch_replacement = last_branch->left;
            last_branch_dangling_child = NULL;
        }

        if (last_branch_parent == NULL)
            tst->head[*key] = last_branch_replacement;
        else {
            if (last_branch_parent->left == last_branch)
                last_branch_parent->left = last_branch_replacement;
            else if (last_branch_parent->right == last_branch)
                last_branch_parent->right = last_branch_replacement;
            else
                last_branch_parent->middle = last_branch_replacement;
        }

        if (last_branch_dangling_child != NULL) {
            current_node = last_branch_replacement;
            while (current_node->left != NULL)
                current_node = current_node->left;
            current_node->left = last_branch_dangling_child;
        }

        next_node = last_branch;
    }

    do {
        current_node = next_node;
        next_node = current_node->middle;

        current_node->left = NULL;
        current_node->right = NULL;
        current_node->middle = tst->free_list;
        tst->free_list = current_node;
    } while (current_node->value != 0);

    return next_node;
}


/*
**  tst_cleanup frees all memory allocated to nodes, internal structures,
**  as well as tst itself.
*/
void
tst_cleanup(struct tst *tst)
{
    struct node_lines *current_line;
    struct node_lines *next_line;

    next_line = tst->node_lines;
    do {
        current_line = next_line;
        next_line = current_line->next;
        free(current_line->node_line);
        free(current_line);
    } while (next_line != NULL);

    free(tst);
}
